### DISCUSSION TITLE
How to signal a parse error from a value callback?

### DISCUSSION THREAD

**Discussion by Korsar13** [foonathan/lexy#147] [Q&A] [answered]
```
std::set<std::string> block_names;

static constexpr auto ident = dsl::identifier( dsl::ascii::alpha, dsl::ascii::alpha_digit_underscore );

struct block_name
{
	static constexpr auto rule  = ident;
	static constexpr auto value = lexy::callback<std::string>(
		[]( const auto& lx ){
			auto s = std::string{ lx.begin(), lx.end() };
			if ( !block_names.insert( s ).second )
			{
				// error "redefinition" ?
			}
			return std::move(s);
		}
	);
};
```

**Comment by foonathan** [comment, marked as answer]
You can't signal a parse error from a callback because those aren't parse errors - they're semantic errors. You probably want to format into your own diagnostics anyway, so can populate that in there. An example of that can be found here: https://github.com/foonathan/lauf/blob/752b53df4b3f6c9affeb300f35bf9a28271a6849/src/lauf/frontend/text.cpp#L253 Or you just generate an AST and do semantic errors later.

**Reply by vobloeb** [reply]
Key uniqueness are indeed parse errors, it can be easily seen in a smaller space, e.g. with 3 possible keys we can construct a simple parser rule
```cpp
rule = LEXY_LIT("a") >> dsl::opt(LEXY_LIT("b") >> dsl::opt(LEXY_LIT("c")) | LEXY_LIT("c") >> dsl::opt(LEXY_LIT("b")))
     | LEXY_LIT("b") >> dsl::opt(LEXY_LIT("a") >> dsl::opt(LEXY_LIT("c")) | LEXY_LIT("c") >> dsl::opt(LEXY_LIT("a")))
     | LEXY_LIT("c") >> dsl::opt(LEXY_LIT("a") >> dsl::opt(LEXY_LIT("b")) | LEXY_LIT("b") >> dsl::opt(LEXY_LIT("a")));
```
and you do something like this parsing integers.

but it quickly became unpractical: parsing `IPv6` is a bit more complex so you resorted to the manual scanner, then `doubles`, then key uniqueness.

The must be a way to write a non-pure validator somewhere between `rule` and `value`, like `.filter` or `.flatMap` in Scala's parser combinators
